src/hardware/voodoo.cpp

```diff
@@ -0,0 +1,7175 @@
+#define RGB_GREEN(rgb)		(((rgb) >> 8) & 0xff)
+#define RGB_BLUE(rgb)		((rgb) & 0xff)
+
+/* common colors */
+#define RGB_BLACK			(MAKE_ARGB(255,0,0,0))
+#define RGB_WHITE			(MAKE_ARGB(255,255,255,255))
+
+/***************************************************************************
+    inline FUNCTIONS
+***************************************************************************/
+
+/*-------------------------------------------------
+    pal5bit - convert a 5-bit value to 8 bits
+-------------------------------------------------*/
+inline UINT8 pal5bit(UINT8 bits)
+{
+	bits &= 0x1f;
+	return (bits << 3) | (bits >> 2);
+}
+
+#ifdef C_ENABLE_VOODOO_DEBUG
+/* rectangles describe a bitmap portion */
+struct rectangle
+{
+	int				min_x;			/* minimum X, or left coordinate */
+	int				max_x;			/* maximum X, or right coordinate (inclusive) */
+	int				min_y;			/* minimum Y, or top coordinate */
+	int				max_y;			/* maximum Y, or bottom coordinate (inclusive) */
+};
+#endif
+
+/* Standard MIN/MAX macros */
+#ifndef MIN
+#define MIN(x,y)			((x) < (y) ? (x) : (y))
+#endif
+#ifndef MAX
+#define MAX(x,y)			((x) > (y) ? (x) : (y))
+#endif
+
+/* Macros for normalizing data into big or little endian formats */
+#define FLIPENDIAN_INT16(x)	(((((UINT16) (x)) >> 8) | ((x) << 8)) & 0xffff)
+#define FLIPENDIAN_INT32(x)	((((UINT32) (x)) << 24) | (((UINT32) (x)) >> 24) | \
+	(( ((UINT32) (x)) & 0x0000ff00) << 8) | (( ((UINT32) (x)) & 0x00ff0000) >> 8))
+#define FLIPENDIAN_INT64(x)	\
+	(												\
+		(((((UINT64) (x)) >> 56) & ((UINT64) 0xFF)) <<  0)	|	\
+		(((((UINT64) (x)) >> 48) & ((UINT64) 0xFF)) <<  8)	|	\
+		(((((UINT64) (x)) >> 40) & ((UINT64) 0xFF)) << 16)	|	\
+		(((((UINT64) (x)) >> 32) & ((UINT64) 0xFF)) << 24)	|	\
+		(((((UINT64) (x)) >> 24) & ((UINT64) 0xFF)) << 32)	|	\
+		(((((UINT64) (x)) >> 16) & ((UINT64) 0xFF)) << 40)	|	\
+		(((((UINT64) (x)) >>  8) & ((UINT64) 0xFF)) << 48)	|	\
+		(((((UINT64) (x)) >>  0) & ((UINT64) 0xFF)) << 56)		\
+	)
+
+#define ACCESSING_BITS_0_15				((mem_mask & 0x0000ffff) != 0)
+#define ACCESSING_BITS_16_31			((mem_mask & 0xffff0000) != 0)
+
+//// constants for expression endianness
+#ifndef WORDS_BIGENDIAN
+#define NATIVE_ENDIAN_VALUE_LE_BE(leval,beval)	(leval)
+#else
+#define NATIVE_ENDIAN_VALUE_LE_BE(leval,beval)	(beval)
+#endif
+
+#define BYTE4_XOR_LE(a) 				((a) ^ NATIVE_ENDIAN_VALUE_LE_BE(0,3))
+
+#define BYTE_XOR_LE(a)  				((a) ^ NATIVE_ENDIAN_VALUE_LE_BE(0,1))
+
+#define profiler_mark_start(x)	do { } while (0)
+#define profiler_mark_end()		do { } while (0)
+
+/* Highly useful macro for compile-time knowledge of an array size */
+#define ARRAY_LENGTH(x)		(sizeof(x) / sizeof(x[0]))
+
+inline INT32 mul_32x32_shift(INT32 a, INT32 b, INT8 shift)
+{
+	return (INT32)(((INT64)a * (INT64)b) >> shift);
+}
+
+#if !defined(__SSE2__) && (_M_IX86_FP == 2 || (defined(_M_AMD64) || defined(_M_X64)))
+#define __SSE2__ 1
+#endif
+#if defined(__SSE2__) && __SSE2__
+#include <emmintrin.h>
+static INT16 sse2_scale_table[256][8];
+#endif
```

Comment:
I'm not even reading these... demacro everything!

Reply:
We'd want to actually performance test without and with the manual SSE optimisations. Maybe the compilers are now clever enough to do all that for us.

Reply:
Compilers definitely are - that's why we upped the minimal requirements to a certain version of Windows, as it enabled more auto-vectorization on MSVC builds.

But it's fragile :( In C++ it's extremely easy to accidentally disable auto-vectorisation... This would be so much easier in Rust :(

Reply:
Another note, partly for myself:

- We want plain, readable version of SSE2 code that is concerned with correctness only and not with speed (for unit testing)
- SSE2 present in this code should be preserved, of course
- … but we should use [function multiversioning](https://gcc.gnu.org/onlinedocs/gcc-9.1.0/gcc/Function-Multiversioning.html) (supported on GCC and Clang, I have no idea about MSVC) to support other instruction sets, [Neon](https://www.arm.com/technologies/neon) (for our RPi and macOS ARM users).

Reply:
(closing for now)

Reply:
> But it's fragile :( In C++ it's extremely easy to accidentally disable auto-vectorisation... This would be so much easier in Rust :(

For autovectorizing info from the compiler:
```
meson setup ... -Dwarning_level=0 -Dautovec_info=true
```

works with GCC and Clang - both giving very helpful and slightly different feedback.

Reply:
Google's Highway is a very flexible and elegant template SIMD library - https://github.com/google/highway. Supports 17 CPU platforms (and more).